File: projects/01-url-shortener/src/shutdown.rs
//! Graceful shutdown: the signal future and the bounded final flush of the
//! click ingestor. Kept out of `main.rs` (wiring only) so the timeout policy —
//! and the tests that pin it — live next to the logic they cover.

use std::time::Duration;

use tracing::{info, warn};

/// How long we wait for the click buffer to flush on shutdown before giving up.
/// Kept comfortably under a typical orchestrator SIGTERM→SIGKILL grace period
/// (k8s default 30s) so we exit cleanly rather than being killed mid-write.
pub(crate) const SHUTDOWN_FLUSH_BUDGET: Duration = Duration::from_secs(5);

/// Result of waiting for the ingestor to flush on shutdown.
#[derive(Debug, PartialEq, Eq)]
pub(crate) enum ShutdownOutcome {
    /// Buffer flushed and the ingestor task returned within budget.
    Flushed,
    /// The ingestor task panicked while flushing (batch may be lost).
    Panicked,
    /// The flush didn't finish within `budget`; we exit without it.
    TimedOut,
}

/// Wait for the ingestor's final flush, bounded by `budget`.
///
/// The channel closes once every `ClickSink` is dropped, which drives the
/// ingestor's `run` loop to flush what's left and return. We join that task so
/// the process doesn't exit mid-flush. The timeout is the accepted tradeoff: a
/// wedged DB write must not hold us past the SIGKILL deadline, so we'd rather
/// drop a final batch than hang.
pub(crate) async fn drain_ingestor(
    handle: tokio::task::JoinHandle<()>,
    budget: Duration,
) -> ShutdownOutcome {
    match tokio::time::timeout(budget, handle).await {
        Ok(Ok(())) => {
            info!("click buffer flushed, ingestor stopped cleanly");
            ShutdownOutcome::Flushed
        }
        Ok(Err(e)) => {
            warn!(error = %e, "ingestor task panicked during shutdown flush");
            ShutdownOutcome::Panicked
        }
        Err(_) => {
            warn!(
                budget_secs = budget.as_secs(),
                "ingestor flush exceeded shutdown budget; exiting with clicks possibly unflushed"
            );
            ShutdownOutcome::TimedOut
        }
    }
}

pub(crate) async fn shutdown_signal() {
    let ctrl_c = async {
        tokio::signal::ctrl_c()
            .await
            .expect("failed to install Ctrl-C handler");
    };

    #[cfg(unix)]
    let terminate = async {
        use tokio::signal::unix::{signal, SignalKind};
        signal(SignalKind::terminate())
            .expect("failed to install term signal handler")
            .recv()
            .await;
    };

    #[cfg(not(unix))]
    let terminate = std::future::pending();

    tokio::select! {
        _ = ctrl_c => {},
        _ = terminate => {},
    }

    info!("shutdown signal received");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn drain_ingestor_reports_a_clean_flush() {
        // A task that returns promptly stands in for the ingestor finishing its
        // final flush well inside the budget.
        let handle = tokio::spawn(async {});
        let outcome = drain_ingestor(handle, Duration::from_secs(5)).await;
        assert_eq!(outcome, ShutdownOutcome::Flushed);
    }

    #[tokio::test]
    async fn drain_ingestor_gives_up_after_budget() {
        // The ingestor is wedged (e.g. a stuck DB write). We must return near the
        // budget, not block on the task: a 50ms budget against a 30s task proves
        // we don't await it.
        let handle = tokio::spawn(async {
            tokio::time::sleep(Duration::from_secs(30)).await;
        });
        let started = std::time::Instant::now();
        let outcome = drain_ingestor(handle, Duration::from_millis(50)).await;
        assert_eq!(outcome, ShutdownOutcome::TimedOut);
        assert!(
            started.elapsed() < Duration::from_secs(5),
            "returned near the budget, not after the 30s task"
        );
    }

    #[tokio::test]
    async fn drain_ingestor_survives_a_panicking_flush() {
        // A panic in the flush must not take the shutdown path down with it.
        let handle = tokio::spawn(async { panic!("simulated flush failure") });
        let outcome = drain_ingestor(handle, Duration::from_secs(5)).await;
        assert_eq!(outcome, ShutdownOutcome::Panicked);
    }
}

File: projects/01-url-shortener/src/url_shortener/shutdown.py
"""Graceful shutdown: the bounded final flush of the click ingestor.

Kept out of `main.py` (wiring only) so the timeout policy - and the tests that
pin it - live next to the logic they cover.

There is no signal handling here, and that is not an omission. Uvicorn installs
the SIGTERM/SIGINT handlers, stops accepting connections, lets in-flight requests
finish, and *then* runs the lifespan's shutdown half. So "drain on SIGTERM" is
implemented by putting this in the lifespan's `finally` - the same place the
pool and the Redis client are closed.
"""

from __future__ import annotations

import asyncio
from enum import Enum
from typing import Final

import structlog

from .ingest import ClickIngestor

__all__ = ["SHUTDOWN_FLUSH_BUDGET", "ShutdownOutcome", "drain_ingestor"]

log = structlog.get_logger(__name__)

SHUTDOWN_FLUSH_BUDGET: Final[float] = 5.0
"""Seconds to wait for the click buffer to flush before giving up.

Comfortably under a typical orchestrator's SIGTERM->SIGKILL grace period (30s on
Kubernetes) so we exit on our own terms rather than being killed mid-write."""


class ShutdownOutcome(Enum):
    """How the final flush went."""

    FLUSHED = "flushed"
    """Buffer written and the ingestor task returned within budget."""

    FAILED = "failed"
    """The ingestor task raised while flushing (the batch may be lost)."""

    TIMED_OUT = "timed_out"
    """The flush did not finish within budget; we exit without it."""


async def drain_ingestor(
    ingestor: ClickIngestor,
    task: asyncio.Task[None],
    budget: float = SHUTDOWN_FLUSH_BUDGET,
) -> ShutdownOutcome:
    """Stop the ingestor and wait for its final flush, bounded by `budget`.

    The budget is the accepted trade-off: a wedged database write must not hold
    the process past the SIGKILL deadline, so we would rather drop one last
    batch of analytics than hang and be killed mid-flush.
    """
    try:
        async with asyncio.timeout(budget):
            await ingestor.stop()
            await task
    except TimeoutError:
        log.warning(
            "ingestor flush exceeded shutdown budget; exiting with clicks possibly unflushed",
            budget_secs=budget,
        )
        task.cancel()
        return ShutdownOutcome.TIMED_OUT
    except Exception as exc:  # noqa: BLE001 - a broken flush must not break shutdown
        log.warning("ingestor task failed during shutdown flush", error=str(exc))
        return ShutdownOutcome.FAILED

    log.info("click buffer flushed, ingestor stopped cleanly")
    return ShutdownOutcome.FLUSHED
